Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift

```diff
@@ -35,83 +126,28 @@ public struct AbstractInterpreter {
         switch instr.op {
-        case is BeginAnyFunctionDefinition:
-            stack.append(currentState)
-        case is EndAnyFunctionDefinition:
-            let functionState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(functionState, previousState))
-            typeChanges += getStateChanges(oldState: functionState, newState: stack.last!)
         case is BeginIf:
-            stack.append(currentState)
+            state.pushNewState()
         case is BeginElse:
-            let ifState = stack.removeLast()
-            ifStack.append(ifState)
-            typeChanges += getStateChanges(oldState: ifState, newState: stack.last!)
+            state.saveIfState()
+            // Exactly one of if/else block is executed. So we should reset types from if block now.
+            state.mergeStates(typeChanges: &typeChanges, reset: true)
         case is EndIf:
-            let ifState = ifStack.removeLast()
-            let elseState = stack.removeLast()
-            stack.append(merge(ifState, elseState))
-            typeChanges += getStateChanges(oldState: elseState, newState: stack.last!)
-        case is BeginWhile:
-            stack.append(currentState)
-        case is EndWhile:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
-        case is BeginDoWhile:
-            stack.append(currentState)
-        case is EndDoWhile:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
-        case is BeginFor:
-            stack.append(currentState)
-        case is EndFor:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
-        case is BeginForIn:
-            stack.append(currentState)
-        case is EndForIn:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
-        case is BeginForOf:
-            stack.append(currentState)
-        case is EndForOf:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
+            state.mergeStates(from: .conditional, typeChanges: &typeChanges)
```

Comment:
Don't you also need to merge with current (main) state here somehow? AFAIU, this wouldn't be handled correctly now:
```
let v0 = 41;
let v1 = 42;
if (x) {
    v0 = "foo";
} else {
    v1 = "bar";
}
```
Because you merge the current state with the state after the if (which I think would just be "v0 is .string"), so you'd miss the change of `v1` at the `EndIf`? Maybe?

Reply:
added test for this